src/index.ts

```diff
@@ -511,120 +537,166 @@ export const plugin: Plugin<[FlexibleContainerOptions?], Root> = (options) => {
   }
 
   const transformer: Transformer<Root> = (tree) => {
-    // if a html node.value ends with "\n:::", remove and carry it into a new paragraph
+    // Do not pre-split paragraphs; handle opening/closing within the analyzer
+    // to preserve expected behavior for invalid patterns.
+
+    // Targeted split: if a paragraph starts with a fence and contains a pure closing
+    // fence line later followed by more content, split into two paragraphs at that line.
+    visit(tree, "paragraph", function (node, index, parent) {
+      /* v8 ignore next */
+      if (!parent || typeof index === "undefined") return;
+      if (node.children.length !== 1 || node.children[0].type !== "text") return;
+
+      const text = (node.children[0] as Text).value;
+      const open = text.match(/^(:{3,})/u);
+      if (!open) return;
+      if (!text.includes("\n")) return;
+
+      const fence = open[1];
+      const lines = text.split(/\n/);
+      // Find a line that equals the same-length fence (allow surrounding spaces) and has content after it
+      let closeIdx = -1;
+      for (let i = 1; i < lines.length; i++) {
+        if (new RegExp(`^\\s*${fence}\\s*$`, "u").test(lines[i])) {
+          closeIdx = i;
+          break;
+        }
+      }
+      if (closeIdx === -1) return;
+      if (closeIdx >= lines.length - 1) return; // nothing after closing -> don't split
+
+      const firstPart = lines.slice(0, closeIdx + 1).join("\n");
+      const secondPart = lines.slice(closeIdx + 1).join("\n");
+
+      const p1 = u("paragraph", [u("text", firstPart)]) as Paragraph;
+      const p2 = u("paragraph", [u("text", secondPart)]) as Paragraph;
+      parent.children.splice(index, 1, p1, p2);
+      return index + 2;
+    });
+
+    // if a html node.value ends with a closing fence, remove and carry it into a new paragraph
     visit(tree, "html", function (node, index, parent) {
       /* v8 ignore next */
       if (!parent || typeof index === "undefined") return;
 
-      if (!/\n:::$/.test(node.value)) return;
+      const m = node.value.match(/\n(:{3,})\s*$/u);
+      if (!m) return;
 
-      node.value = node.value.replace(/\n:::$/, "");
+      node.value = node.value.replace(/\n(:{3,})\s*$/u, "");
 
-      const p = u("paragraph", [u("text", "\n:::")]);
+      const p = u("paragraph", [u("text", "\n" + m[1])]);
 
       // add the paragraph after the html node, in order to the next visitor can catch the container node
       parent.children.splice(index + 1, 0, p);
     });
 
-    // main visit
-    visit(tree, "paragraph", function (node, index, parent) {
-      /* v8 ignore next */
-      if (!parent || typeof index === "undefined") return;
+    // multi-pass visit to allow nested containers
+    for (let pass = 0; pass < 10; pass++) {
```

Comment:
ten times visit ? why ten, it is not feasible, for nesting it is not necessary. The visit function can be recursive if you return a specific index.